lookup_optim.go

```diff
@@ -0,0 +1,311 @@
+package dht
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/libp2p/go-libp2p-kad-dht/metrics"
+	"github.com/libp2p/go-libp2p-kad-dht/netsize"
+	"github.com/libp2p/go-libp2p-kad-dht/qpeerset"
+	kb "github.com/libp2p/go-libp2p-kbucket"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/multiformats/go-multihash"
+	ks "github.com/whyrusleeping/go-keyspace"
+	"gonum.org/v1/gonum/mathext"
+)
+
+const (
+	// OptProvIndividualThresholdCertainty describes how sure we want to be that an individual peer that
+	// we find during walking the DHT actually belongs to the k-closest peers based on the current network size
+	// estimation.
+	OptProvIndividualThresholdCertainty = 0.9
+
+	// OptProvSetThresholdStrictness describes the probability that the set of closest peers is actually further
+	// away then the calculated set threshold. Put differently, what is the probability that we are too strict and
+	// don't terminate the process early because we can't find any closer peers.
+	OptProvSetThresholdStrictness = 0.1
+
+	// OptProvReturnRatio corresponds to how many ADD_PROVIDER RPCs must have completed (regardless of success)
+	// before we return to the user. The ratio of 0.75 equals 15 RPC as it is based on the Kademlia bucket size.
+	OptProvReturnRatio = 0.75
+)
+
+type addProviderRPCState int
+
+const (
+	Sent addProviderRPCState = iota + 1
+	Success
+	Failure
+)
```

Comment:
Do these need to be exported? If not, might be a good idea to unexport

Reply:
of course, they don't need to be exported, thanks!